src/couch_epi_data_source.erl

```diff
@@ -0,0 +1,173 @@
+    Interval = proplists:get_value(interval, Options, ?MONITOR_INTERVAL),
+    {ok, _Timer} = timer:send_interval(Interval, self(), tick),
+    {ok, #state{
+        subscriber = Subscriber,
+        locator = Locator,
+        key = Key,
+        handle = couch_epi_data_gen:get_handle(Key)}}.
+
+handle_call(wait, _From, #state{initialized = true} = State) ->
+    {reply, ok, State};
+handle_call(wait, From, #state{pending = Pending} = State) ->
+    {noreply, State#state{pending = [From | Pending]}};
+handle_call(reload, _From, State) ->
+    {Res, NewState} = reload_if_updated(State),
+    {reply, Res, NewState};
+handle_call(_Request, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast(init, #state{pending = Pending} = State) ->
+    {_, NewState} = reload_if_updated(State),
+    [gen_server:reply(Client, ok) || Client <- Pending],
+    {noreply, NewState#state{initialized = true, pending = []}};
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(tick, State0) ->
+    {_Res, State1} = reload_if_updated(State0),
+    {noreply, State1};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+
+locate(App, {priv_file, FileName}) ->
+    case priv_path(App, FileName) of
+        {ok, FilePath} ->
+            ok = ensure_exists(FilePath),
+            {ok, {file, FilePath}};
+        Else ->
+            Else
+    end;
+locate(_App, {file, FilePath}) ->
+    ok = ensure_exists(FilePath),
+    {ok, {file, FilePath}}.
+
+priv_path(AppName, FileName) ->
+    case code:priv_dir(AppName) of
+        {error, _Error} = Error ->
+            Error;
+        Dir ->
+            {ok, filename:join(Dir, FileName)}
+    end.
+
+ensure_exists(FilePath) ->
+    case filelib:is_regular(FilePath) of
+        true ->
+            ok;
+        false ->
+            {error, {notfound, FilePath}}
+    end.
+
+reload_if_updated(#state{hash = OldHash, locator = Locator} = State) ->
+    case read(Locator) of
+        {ok, OldHash, _Data} ->
+            {ok, State};
+        {ok, Hash, Data} ->
+            safe_set(Hash, Data, State);
+        Else ->
+            {Else, State}
+    end.
+
+safe_set(Hash, Data, #state{} = State) ->
+    #state{
+        handle = Handle,
+        subscriber = Subscriber,
+        key = Key} = State,
+
+    try
+        OldData = current(Handle, Subscriber),
+        ok = couch_epi_data_gen:set(Handle, Subscriber, Data),
+        couch_epi_server:notify(Subscriber, Key, {data, OldData}, {data, Data}),
+        {ok, State#state{hash = Hash}}
+    catch Class:Reason ->
+        {{Class, Reason}, State}
+    end.
+
+read({file, FilePath}) ->
+    case file:consult(FilePath) of
+        {ok, Data} ->
+            {ok, hash(FilePath), Data};
```

Comment:
Hm. Actually hash(FilePath) is correct. Since the hash function would read the file.